### ISSUE TITLE
[RLlib] Dreamer seems to try to predict the previous reward

### ISSUE THREAD

**Issue by krisbrud** [ray-project/ray#31009]
### What happened + What you expected to happen

Hi!

I'm currently working on using the Ray RLlib Dreamer implementation with a custom environment, and enjoy the framework a lot so far. Nevertheless, I have run into something in the implementation that may be unintended.

In [`DreamerTorchPolicy.postprocess_trajectory()`](https://github.com/ray-project/ray/blob/66373e03211dd56d325456dac10867e46fccce38/rllib/algorithms/dreamer/dreamer_torch_policy.py#L172-L174), we find this comment (and an implementation that is done accordingly).

```
"""Batch format should be in the form of (s_t, a_(t-1), r_(t-1))
When t=0, the resetted obs is paired with action and reward of 0.
"""
```

Here, it is very unclear for me why the batch format should not be of the form `(s_t, a_(t-1), r_t)`, i.e. using the current reward (`r_t`) rather than the one from the previous timestep (`r_(t-1)`).

In the original paper [Dream to Control](https://arxiv.org/abs/1912.01603), the reward model (equation 1) is `q(r_t | s_t)`, where `s_t` is the latent state. The posterior/representation model `p(s_t | s_(t-1), a_(t-1), o_t)` is used to estimate the current state `s_t` based on the last state-action pair `(s_(t-1), a_(t-1))` as well as the current observation `o_t`.

For reference, `s_t` is to my understanding implemented through the `features` variable with the following implementation in the `loss()` function, which is used to predict rewards (`reward_pred`):
```
latent = self.model.encoder(train_batch["obs"])
post, prior = self.model.dynamics.observe(latent, train_batch["actions"])
features = self.model.dynamics.get_feature(post)
image_pred = self.model.decoder(features)
reward_pred = self.model.reward(features)
image_loss = -torch.mean(image_pred.log_prob(train_batch["obs"]))
reward_loss = -torch.mean(reward_pred.log_prob(train_batch["rewards"]))
```

The paper also shows this figure (figure 3) showing the intended structure of the model, 
<img width="367" alt="image" src="https://user-images.githubusercontent.com/32980192/206853204-08de2de3-1e26-46f9-872a-1e33c5065083.png">

I believe that the implementation may in practice look more like this (excuse me for the quick and dirty graphics): 
![image](https://user-images.githubusercontent.com/32980192/206853492-6cea8cb2-e001-4b86-aca5-b3127bd93d9b.png)

My hypothesis for why it may still seem to work is that the model has a decently sized latent space, it is possible that it is able to retain features that may be used to decode the previous reward.

It seems that @michaelzhiluo and @ArturNiederfahrenhorst have worked on this, is this is the intended behaviour? 

Other than this I just want to say that I find the implementation very readable, much clearer than other implementations I have looked into :).

### Versions / Dependencies

`ray==2.1.0`

### Reproduction script

No reproduction script, but here is the current implementation of the method:

```
@override(TorchPolicyV2)
def postprocess_trajectory(
    self,
    sample_batch: SampleBatch,
    other_agent_batches: Optional[
        Dict[AgentID, Tuple["Policy", SampleBatch]]
    ] = None,
    episode: Optional["Episode"] = None,
) -> SampleBatch:
    """Batch format should be in the form of (s_t, a_(t-1), r_(t-1))
    When t=0, the resetted obs is paired with action and reward of 0.
    """
    obs = sample_batch[SampleBatch.OBS]
    new_obs = sample_batch[SampleBatch.NEXT_OBS]
    action = sample_batch[SampleBatch.ACTIONS]
    reward = sample_batch[SampleBatch.REWARDS]
    eps_ids = sample_batch[SampleBatch.EPS_ID]

    act_shape = action.shape
    act_reset = np.array([0.0] * act_shape[-1])[None]
    rew_reset = np.array(0.0)[None]
    obs_end = np.array(new_obs[act_shape[0] - 1])[None]

    batch_obs = np.concatenate([obs, obs_end], axis=0)
    batch_action = np.concatenate([act_reset, action], axis=0)
    batch_rew = np.concatenate([rew_reset, reward], axis=0)
    batch_eps_ids = np.concatenate([eps_ids, eps_ids[-1:]], axis=0)

    new_batch = {
        SampleBatch.OBS: batch_obs,
        SampleBatch.REWARDS: batch_rew,
        SampleBatch.ACTIONS: batch_action,
        SampleBatch.EPS_ID: batch_eps_ids,
    }
    return SampleBatch(new_batch)
```

### Issue Severity

Low: It annoys or frustrates me.

**Comment by ArturNiederfahrenhorst** [comment]
Hi @krisbrud , thanks for filing this issue and glad you enjoy RLlib.
The comment you refer to is not very clear, especially since we have two keys in SampleBatch that refer to `s_t`.
SampleBatch.OBS and SampleBatch.NEXT_OBS. 
In accordance with that in the code below and the original paper, a column of a batch should contain:

- observation o at t
- action that led to o (action picked between t and t-1)
- reward received together with o

**Comment by ArturNiederfahrenhorst** [comment]
Can you spot anything in the code outside this comment that points to this also being algorithmically wrong?

**Comment by krisbrud** [comment]
Hi! Thanks for the response @ArturNiederfahrenhorst. You are correct, and I think I have located the source of my confusion.

Let's take a step back and look at how the OpenAI gym API is presented in the [original paper](https://arxiv.org/pdf/1606.01540.pdf) (end of the background section):
```
ob0 = env.reset() # sample environment state, return first observation
a0 = agent.act(ob0) # agent chooses first action
ob1, rew0, done0, info0 = env.step(a0) # environment returns observation,
# reward, and boolean flag indicating if the episode is complete.
```

According to my understanding, we want the posterior `post[t]` to depend on the observation `o[t]` as well as the action leading up to it, `a[t-1]`. Therefore we want to use the posterior `post[t]` to predict the reward `r[t-1]`, as `r[t-1]` inherently depends on `a[t-1]`.

The current implementation looks like this to me for an episode with `T` actions taken, as well as the initial observation that resetting the environment gives:
```
batch_obs     = [o[0], o[1], o[2], ..., o[T-1], obs_end]
batch_actions = [0,    a[0], a[1], ..., a[T-2], a[T-1]]
batch_rewards = [0,    r[0], r[1], ..., r[T-2], r[T-1]]
```
Where `obs_end` is the new observation returned when taking the last action `a[T-1]`. With the reasoning above, I guess the implementation should be correct.

## So what is the source of my confusion? 
With Sutton & Barto's definition (disregarding the difference between the state and observation), we have that taking `action[t]` in `state[t]` gives `obs[t+1], reward[t+1]` (instead of `reward[t]` as in the `gym` API!).  I guess I have seen this figure reiterated many enough times that I forgot that this is not how it is implemented in `gym`. I'm probably not the first person that has been confused by this either.

![image](https://user-images.githubusercontent.com/32980192/207864496-b3111dd1-7c81-418e-87f2-78f7482f0008.png)

Nevertheless, it would be great if the RLlib convention for aligning the samples (or at least what `RolloutWorker` does) would be documented. I guess it doesn't affect people using the implemented algorithms that much, but it would be helpful for modifying existing algorithms or implementing new ones. Maybe it is documented somewhere, but I can't seem to find it.

**Comment by ArturNiederfahrenhorst** [comment]
Thanks for the in-depth analysis.
I agree - the definition is not apparent and, thus, should be stated somewhere.
We are looking into SampleBatch, but I think that is where it should be documented for the most part.
I'll include that in my PR!

**Comment by ArturNiederfahrenhorst** [comment]
I have included our internal definitions in the SampleBatch documentation.
https://github.com/ray-project/ray/pull/31101/files
Thanks for raising this again!

**Comment by krisbrud** [comment]
Okay, great! Sounds good. Thanks for the help:)
